aggregator/src/aggregator/queue.rs

```diff
@@ -0,0 +1,764 @@
+                                    }
+                                } else {
+                                    permit_tx.send(Err(Error::TooManyRequests));
+                                }
+                            },
+                            DispatcherMessage::Cancel(id) => {
+                                debug!(?id, "removing from the queue");
+                                stack.remove(&id);
+                            },
+                        }
+                    }
+
+                    Ok(permit) = semaphore.acquire_owned() => {
+                        // pop_last() pops the element maximum valued element. The request ID
+                        // counter is _mostly_ monotonically incrementing, so the highest ID will
+                        // be the most recent request, giving us LIFO semantics.
+                        //
+                        // This property isn't preserved when the ID generator overflows, but that
+                        // is not a practical concern. See [`Self::id_counter`].
+                        match stack.pop_last() {
+                            Some((_, permit_tx)) => permit_tx.send(Ok(permit)),
+                            None => permits.merge(permit),
+                        }
+                    }
+                }
+
+                // Unwrap safety: only fails on architectures where usize is less than 32 bits, or
+                // greater than 64 bits.
+                metrics.outstanding_requests.record(
+                    (stack.len() + usize::try_from(concurrency).unwrap() - permits.num_permits())
+                        .try_into()
+                        .unwrap(),
+                    &[],
+                );
+            }
+        })
+    }
+
+    async fn acquire(&self) -> Result<OwnedSemaphorePermit, Error> {
+        let id = self.id_counter.fetch_add(1, Ordering::Relaxed);
+        let (permit_tx, permit_rx) = oneshot::channel();
+
+        self.dispatcher_tx
+            .send(DispatcherMessage::Enqueue(id, PermitTx(permit_tx)))
+            // We don't necessarily panic because the dispatcher task could be shutdown as part of
+            // process shutdown, while a request is in flight.
+            .map_err(|_| Error::Internal("dispatcher task died".to_string()))?;
+
+        let mut drop_guard = CancelDropGuard {
+            id,
+            cancel_tx: self.dispatcher_tx.clone(),
+            done: false,
+        };
+        let permit = permit_rx.await;
+        drop_guard.done = true;
+
+        // If the rx channel is prematurely dropped, we'll reach this error, indicating that
+        // something has gone wrong with the dispatcher task. If the drop guard causes the rx
+        // channel to be dropped, we shouldn't reach this error because the overall future would
+        // have been dropped.
+        permit.map_err(|_| Error::Internal("rx channel unexpectedly dropped".to_string()))?
+    }
+}
+
+impl Drop for LIFORequestQueue {
```

Comment:
I'm fine with taking this change, agreed with your rationale.

How should we service requests during process shutdown? That is, should we drop them on the floor with 500 Internal Server Error, or block the process shutdown while they're serviced normally?

I went with returning immediately when `.recv()` shows the channel is closed. This will drop the queue without flushing it, and any waiting requests will error out because their oneshot senders will have been closed.

I think this is the easiest approach, for now. Otherwise we would need to block process shutdown on the dispatcher working through its queue, which is a fair bit of plumbing I think (and requires tracking a join handle anyways). But whether we should do that depends on what we want to do with requests during process shutdown.

Reply:
(FWIW I think we have a broader problem of not fully waiting for all important background tasks shutting down during process shutdown, so I don't think we should necessarily solve it here)

Reply:
> How should we service requests during process shutdown? That is, should we drop them on the floor with 500 Internal Server Error, or block the process shutdown while they're serviced normally?

The typical way to handle process shutdown in a container-orchestration system:
1) Enter a "lame-duck" mode where no further requests are routed to the process, and send a signal to the process to shutdown (via `SIGTERM`, typically).
2) Wait some time to allow the process to handle any final requests & terminate cleanly.
3) If the process hasn't cleanly terminated, hard-kill it (via `SIGKILL`).

We do implement the proper [signal handling logic](https://github.com/divviup/janus/blob/38a671dcd5a6c17f0113cc093490770d845665b6/aggregator/src/binary_utils.rs#L487), and Kubernetes implements roughly the above process when performing rolling restarts.

So -- I think we want to service these requests normally. Dropping requests on the floor during shutdown is pretty rude.


> I think this is the easiest approach, for now. Otherwise we would need to block process shutdown on the dispatcher working through its queue, which is a fair bit of plumbing I think (and requires tracking a join handle anyways). But whether we should do that depends on what we want to do with requests during process shutdown.

I don't think we need to do that. Instead, we can use a conditional select to stop polling the channel once we have reached the end (i.e. received `None`). Then after each loop iteration, if we have reached the end of the channel & the stack is empty, we can exit the loop. (Conditional select docs: see the `, if` examples of `select!` usage in the [rustdocs](https://docs.rs/tokio/latest/tokio/macro.select.html).)

I think this solves termination: we will happily continue servicing requests until we run out of requests to service, or the process is killed.

Reply:
FWIW, I'm unsure that our shutdown handling is correct at the moment -- I need to dig through trillium's code to be sure it doesn't just stop handling existing requests when the `Stopper` is stopped; in non-trillium uses of the `Stopper` we use `stop_future` which simply terminates the future when the stopper is stopped, which definitely isn't what we want for anything that might be handling a request. But that is definitely outside the scope of this PR.

Reply:
I will separately evaluate our shutdown handling for https://github.com/divviup/janus-ops/issues/1994, i.e. make sure we're exhibiting the lame duck behavior you describe, so that we don't drop requests on the floor during scale-down actions.

> edit: ah, you're correct about acquire holding the channel open until the operation is complete.

Yeah I think the latest is straight, though it might be good to have some graceful shutdown tests to evaluate this. We do have graceful shutdown integration tests, but they don't appear to check whether requests are handled properly